sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketStreamSuite.scala

```diff
@@ -0,0 +1,246 @@
+        batchReader.createDataReaderFactories().asScala.map(_.createDataReader()).foreach { r =>
+          while (r.next()) {
+            batch1.append(r.get())
+          }
+        }
+        assert(batch1.map(_.getAs[String](0)) === Seq("hello"))
+        val batch1Stamp = batch1.map(_.getAs[Timestamp](1)).head
+
+        serverThread.enqueue("world")
+        while (batchReader.getEndOffset === offset1) {
+          batchReader.setOffsetRange(Optional.of(offset1), Optional.empty())
+          Thread.sleep(10)
+        }
+        val offset2 = batchReader.getEndOffset
+        val batch2 = new ListBuffer[Row]
+        batchReader.createDataReaderFactories().asScala.map(_.createDataReader()).foreach { r =>
+          while (r.next()) {
+            batch2.append(r.get())
+          }
+        }
+        assert(batch2.map(_.getAs[String](0)) === Seq("world"))
+        val batch2Stamp = batch2.map(_.getAs[Timestamp](1)).head
+        assert(!batch2Stamp.before(batch1Stamp))
+      }
+
+      // Try stopping the source to make sure this does not block forever.
+      batchReader.stop()
+      batchReader = null
+    }
+  }
+
+  test("params not given") {
+    val provider = new TextSocketSourceProvider
+    intercept[AnalysisException] {
+      provider.createMicroBatchReader(Optional.empty(), "",
+        new DataSourceOptions(Map.empty[String, String].asJava))
+    }
+    intercept[AnalysisException] {
+      provider.createMicroBatchReader(Optional.empty(), "",
+        new DataSourceOptions(Map("host" -> "localhost").asJava))
+    }
+    intercept[AnalysisException] {
+      provider.createMicroBatchReader(Optional.empty(), "",
+        new DataSourceOptions(Map("port" -> "1234").asJava))
+    }
+  }
+
+  test("non-boolean includeTimestamp") {
+    val provider = new TextSocketSourceProvider
+    val params = Map("host" -> "localhost", "port" -> "1234", "includeTimestamp" -> "fasle")
+    intercept[AnalysisException] {
+      val a = new DataSourceOptions(params.asJava)
+      provider.createMicroBatchReader(Optional.empty(), "", a)
+    }
+  }
+
+  test("user-specified schema given") {
+    val provider = new TextSocketSourceProvider
+    val userSpecifiedSchema = StructType(
+      StructField("name", StringType) ::
+      StructField("area", StringType) :: Nil)
+    val params = Map("host" -> "localhost", "port" -> "1234")
+    val exception = intercept[AnalysisException] {
+      provider.createMicroBatchReader(
+        Optional.of(userSpecifiedSchema), "", new DataSourceOptions(params.asJava))
+    }
+    assert(exception.getMessage.contains(
+      "socket source does not support a user-specified schema"))
+  }
+
+  test("no server up") {
+    val provider = new TextSocketSourceProvider
+    val parameters = Map("host" -> "localhost", "port" -> "0")
+    intercept[IOException] {
+      batchReader = provider.createMicroBatchReader(
+        Optional.empty(), "", new DataSourceOptions(parameters.asJava))
+    }
+  }
+
```

Comment:
what happened to the input row metrics test?

Reply:
Because of the changes of data source APIs, so for now it is hard to get DF's metrics from V2 API, that's why I deleted this test.

Reply:
the test below seems good replacement to me.